### ISSUE TITLE
Unnecessary forward declaration for type previously used in structure

### ISSUE THREAD

**Issue by osandov** [include-what-you-use/include-what-you-use#970]
The following C source code:

```c
struct foo {
	struct bar *bar;
};

void baz(struct bar *);
```

Gets the following warning from IWYU 0.16:

```

test.c should add these lines:
struct bar;

test.c should remove these lines:

The full include-list for test.c:
struct bar;
---
```

The suggested forward declaration of `struct bar` is unnecessary since its use in `struct foo` has already provided a declaration. Note that neither Clang nor GCC warn for this source file even with `-Wall -Wextra -Wpedantic`. Is this a policy choice or a bug?

(IWYU correctly doesn't warn if I remove the function declaration, or declare a `struct bar *` variable instead of a function.)

**Comment by kimgr** [comment]
This was an explicit choice: https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu.cc#L4064
Clang used to warn with `-Wvisibility` -- not sure if that warning has been removed or moved out of (all, extra, pedantic)... Can you try with that warning explicitly enabled and see what happens?

**Comment by osandov** [comment]
Clang 12 doesn't seem to warn:

```
$ cat test.c
struct foo {
        struct bar *bar;
};

void baz(struct bar *);
$ clang --version
clang version 12.0.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
$ clang -Wall -Wextra -Wvisibility -Werror -c test.c
$ clang -Weverything -Werror -c test.c
$
```

**Comment by kimgr** [comment]
Interesting. Same if you give `baz` a body?

**Comment by osandov** [comment]
Yup, no relevant warnings:

```
$ cat test.c                                
struct foo {
        struct bar *bar;
};

int baz(struct bar *my_bar)
{
        return my_bar == 0;
}
$ clang -Wall -Wextra -Wvisibility -c test.c
$ clang -Weverything -c test.c
test.c:5:5: warning: no previous prototype for function 'baz' [-Wmissing-prototypes]
int baz(struct bar *my_bar)
    ^
test.c:5:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
int baz(struct bar *my_bar)
^
static 
1 warning generated.
```

**Comment by osandov** [comment]
Maybe this was fixed in Clang in llvm/llvm-project@38c202e97e51de7d4f2b523e15fa56c8c1a494cd?

**Comment by kimgr** [comment]
Yes, I think that `if (!PrevDecl)` addition softened Clang's warning a little.

I overlooked the fact that your example has a struct decl inside another struct, it didn't occur to me that that would count as a prev decl. I wonder if we could do a similar check in IWYU to detect that a decl is already visible in the same file even if it's not strictly a forward decl.

**Comment by osandov** [comment]
Yes, that sounds like the right thing to do. This also applies to return values and variable declarations. Basically, any use other than in a parameter list declares the structure in the enclosing scope, so all of these are okay by the C standard:

```c
struct contains_member {
	struct member *member;
};
void use_member(struct member *);

struct returned *create_returned(void);
void use_returned(struct returned *);

extern struct variable var;
void use_variable(struct variable *);
```

But GCC and Clang of course warn for this:

```c
void use_never_seen_before(struct never_seen_before *);
```

**Comment by kimgr** [comment]
For posterity, I tried this, but couldn't get it to work. The `PrevDecl` in `Sema::ActOnTag` is not strictly represented in the decl's redeclaration chain once the AST is fully built. I don't have any obvious ideas for a solution/workaround there at the moment.

**Comment by kimgr** [comment]
IWYU no longer complains about any of these examples on mainline. Closing as fixed.
